### PR TITLE
Hot loading logic for baobab-hot-loader and changed readme

### PR DESCRIPTION
**Pull request by christianalfoni**
Initially it was a bit difficult understanding what "root" and "branch" did in the examples. I cleaned up the examples a bit making them more focused on the pattern with "real life" examples, rather than focusing on these two "root" and "branch" concepts. Think it is easier to understand by focusing on the pattern and just showing code.

Sorry to remove the "point to context" examples, but I do not think developers should do that, unless they really know what they are doing. And if they know what they are doing, they probably know how to point to context ;-)

Also think the Baobab repo should be Baobab stuff, so stuff related specifically to Baobab should be in that repo. This repo should only be about the patterns? Afraid things get a bit mixed. Its all about the patterns really.

Anyways, just a suggestion :-)

### CODE REVIEW

**Comment by ldesplat** [comment]
I must be missing something but aren't the cursors passed through props and not state? At least when I try it with hoc, that's what happens... I see the same issue in your preview article (which is very useful btw).

**Comment by christianalfoni** [comment]
Hi @ldesplat ,

I will expand on a section in the article explaining how state is moved from the tree to the component. Let me double check the different patterns as I have not tested them extensively myself :-)

Thanks for bringing it to my attention!

**Comment by christianalfoni** [comment]
And you are perfectly right @ldesplat , I did not test and read the documentation well enough. I have updated the readme now and will also update the article. Thanks for pointing it out!

**Comment by christianalfoni** [comment]
Okay, I thought it would create a new pull request. So please ignore the README file if that is not suitable for pull. But the rest is "drying" up the code a bit and allowing for hot loading logic. Note that hot loading logic only runs when hot loading is active.

This is part of baobab-hot-loader, testing here: https://www.youtube.com/watch?v=iVYF-_gjJGg

**Comment by christianalfoni** [comment]
Repo is over here: https://github.com/christianalfoni/baobab-hot-loader/blob/master/README.md

**Comment by Yomguithereal** [comment]
@christianalfoni, can you describe what should be done here to enable hot loading?

**Comment by christianalfoni** [comment]
Ah, sorry, that is a pretty important detail :-)

The `makeHot` methods added to `baobab-react` is used by `react-hot-loader` to trigger the extra behaviour needed by the mixins to use hot-loading.

As a developer all you have to do is install `baobab-hot-loader` and add it to your config, here is an example:

```js
var webpackConfig = {
  entry: [
    'webpack/hot/dev-server', 
    path.resolve(appPath, 'main.js')
  ],
  devtool: 'eval-source-map',
  output: {
    filename: 'main.js',
    path: buildPath,
    publicPath: '/build/'
  },
  module: {
    loaders: [{
      test: /\.js$/,
      loaders: 'react-hot!baobab-hot!babel', // Add the loader
      exclude: [node_modules_dir]
    }, {
      test: /\.css$/,
      loader: 'style!css'
    }, {        
      test: /\.less$/,
      loader: 'style!css!less'  
     }]
  },
  plugins: [new Webpack.HotModuleReplacementPlugin()]
};
```

I though I could add that documentation to the `baboab-hot-loader` repo. There are no requirements regarding `baboab-react`, the changes are only related to exposing a method that `baobab-hot-loader` can use.

**Comment by christianalfoni** [comment]
Actually, let me explain how this actually works :-)

1. The file in your project containing the Baobab dependency will be analyzed. The analyzis finds the variable used to set the state of the Baobab tree, which has to be a separate module. For example a `state.js` file containing the state object

2. When the file containing this logic has a "hot update" it will require the new version of the file and set it on the tree.  This is handling changes to the state tree itself

3. The `makeHot` on mixins is related to handling changes to the cursor/facet object inside a component. It needs to clean up the state object in the component and grab fresh data from the tree

**Comment by christianalfoni** [comment]
Regarding point 2. I am now creating an object path diff lib to use with `baobab-hot-loader`. That way it will only update paths in the tree that has actually changed. That way you can safely for example get some data from the server that is put into the tree and when you change som other path in the tree, only that path will update, not reset changes to the tree done by the application itself :-)

**Comment by christianalfoni** [comment]
What do you know, it already exists :-D https://github.com/flitbit/diff

**Comment by Yomguithereal** [comment]
So, what about this PR? Is there something to pull from it?

**Comment by christianalfoni** [comment]
Everything except README, unless you agree with all the changes there ;-)

To summarize:

1. DRYs up the code a bit, now uses a `utils/mixin.js` to run the logic for both mixins and higher-order/decorators

2. Changed `componentDidMount` to `componentWillMount`... not sure why didMount is used?

3. Added a `makeHot` method to mixins and higher-order to allow `baobab-hot-loader` to set the isHot flag

4. `componentWillUpdate` added with a isHot flag to run the cursors/facets property diff

Okay, that should be it!

**Comment by christianalfoni** [comment]
Where did we land on this @Yomguithereal ? :-) It is just too cool and a really cool "selling point" to get people to try Baobab. Not saying Baobab solves everything, but it is important to have something that motivates people to give it a try at least ;-)

**Comment by Yomguithereal** [comment]
Hello @christianalfoni, I need to refocus on it. This is still about compatibility with `react-hot-loader` isn't it? Any news about the upcoming changes about `react-hot-loader` by the way?

File: README.md [hunk 1]
```diff
@@ -42,576 +40,311 @@ Then require the desired pattern and only this one will be loaded (this means th
...
 
-## On root & branches
-
-In order to keep component definitions detached from any particular instance of Baobab, I divided the mixins, higher order components etc. into two:
+## The concept
+All the patterns share two concepts, **root** and **branch**. 
 
-* The **Root** aims at passing a baobab tree through context so that child component (branches) may use it. Typically, your app's top-level component will probably be a root.
-* The **Branches**, bound to cursors, get their data from the tree given by the root.
+* The **root** aims at passing a baobab tree through the context of your React componen tree. That way any child component can use the tree without depending on a parent wrapper and you do not depend on the tree itself when defining a component
```
Comment (tmbtech) [inline README.md]: `componen tree` => `component tree`

**Comment by christianalfoni** [comment]
Hi again, sorry for late replies... I am in real loss of time these days. Jup, this is for compatability with react-hot-loader.

Maybe finish up version 2.0 and create a branch and we can have a "little project" on it? :-)

**Comment by Yomguithereal** [comment]
This is a good idea. I will finish v2 and adapt `baobab-react` to it so we can work out hot-reloading! The v2 version of `baobab-react` should be simpler then since facets have been merged with the cursors API.

**Comment by chcokr** [comment]
Hi, I was wondering how the timeline looks as to baobab-hot-loading starting to work! Looking forward to it :)

**Pull request closed without merging**